src/utils/viewer/ObjectPickingHelper.cpp

```diff
@@ -0,0 +1,132 @@
+#include <Magnum/Shaders/MeshVisualizer.h>
+
+namespace Cr = Corrade;
+namespace Mn = Magnum;
+using Mn::Math::Literals::operator""_rgbf;
+
+ObjectPickingHelper::ObjectPickingHelper(Mn::Vector2i viewportSize) {
+  // create the framebuffer and set the color attachment
+  recreateFramebuffer(viewportSize).mapForDraw();
+  CORRADE_INTERNAL_ASSERT(
+      selectionFramebuffer_.checkStatus(Mn::GL::FramebufferTarget::Draw) ==
+      Mn::GL::Framebuffer::Status::Complete);
+
+  // create a shader
+  Mn::Shaders::MeshVisualizer3D::Flags flags =
+      Mn::Shaders::MeshVisualizer3D::Flag::Wireframe;
+  shader_ = std::make_unique<Mn::Shaders::MeshVisualizer3D>(flags);
+
+  shader_->setViewportSize(Mn::Vector2{viewportSize});
+
+  if (flags & Mn::Shaders::MeshVisualizer3D::Flag::Wireframe) {
+    (*shader_).setColor(0x2f83cc_rgbf).setWireframeColor(0xdcdcdc_rgbf);
+  }
+}
+
+ObjectPickingHelper& ObjectPickingHelper::recreateFramebuffer(
+    Mn::Vector2i viewportSize) {
+  // setup an offscreen frame buffer for object selection
+  selectionDepth_.setStorage(Mn::GL::RenderbufferFormat::DepthComponent24,
+                             viewportSize);
+  selectionDrawableId_.setStorage(Mn::GL::RenderbufferFormat::R32UI,
+                                  viewportSize);
+  selectionFramebuffer_ = Mn::GL::Framebuffer{{{}, viewportSize}};
+  selectionFramebuffer_
+      .attachRenderbuffer(Mn::GL::Framebuffer::BufferAttachment::Depth,
+                          selectionDepth_)
+      .attachRenderbuffer(Mn::GL::Framebuffer::ColorAttachment{1},
+                          selectionDrawableId_);
+  return *this;
+}
+
+ObjectPickingHelper& ObjectPickingHelper::prepareToDraw() {
+  selectionFramebuffer_.bind();
+  mapForDraw();
+  selectionFramebuffer_.clearDepth(1.0f).clearColor(1, Mn::Vector4ui{0xffff});
+  CORRADE_INTERNAL_ASSERT(
+      selectionFramebuffer_.checkStatus(Mn::GL::FramebufferTarget::Draw) ==
+      Mn::GL::Framebuffer::Status::Complete);
+
+  // remove any visualized object that is picked before
+  if (meshVisualizerDrawable_) {
+    delete meshVisualizerDrawable_;
+    meshVisualizerDrawable_ = nullptr;
+  }
+
+  return *this;
+}
+
+ObjectPickingHelper& ObjectPickingHelper::mapForDraw() {
+  selectionFramebuffer_.mapForDraw({{Mn::Shaders::Generic3D::ColorOutput,
+                                     Mn::GL::Framebuffer::DrawAttachment::None},
+                                    {Mn::Shaders::Generic3D::ObjectIdOutput,
+                                     Mn::GL::Framebuffer::ColorAttachment{1}}});
+  return *this;
+}
+
+ObjectPickingHelper& ObjectPickingHelper::handleViewportChange(
+    Mn::Vector2i viewportSize) {
+  recreateFramebuffer(viewportSize);
+  selectionFramebuffer_.setViewport({{}, viewportSize});
+
+  shader_->setViewportSize(Mn::Vector2{viewportSize});
+  return *this;
+}
+
+unsigned int ObjectPickingHelper::getObjectId(
+    Magnum::Vector2i mouseEventPosition,
+    Magnum::Vector2i windowSize) {
```

Comment:
```suggestion
    const Magnum::Vector2i& mouseEventPosition,
    const Magnum::Vector2i& windowSize) {
```

I think it's like this everywhere else, so why not be consistent. Same in other cases.

Reply:
because I thought Vector2i should be very light, almost like an "int", where I thought "passing by value" was OK.